src/Functions/fromReadableSize.cpp

```diff
@@ -0,0 +1,265 @@
+        {"mib", 1048576},               // 1024 * 1024
+        {"gib", 1073741824},            // 1024 * 1024 * 1024
+        {"tib", 1099511627776},         // 1024 * 1024 * 1024 * 1024
+        {"pib", 1125899906842624},      // 1024 * 1024 * 1024 * 1024 * 1024
+        {"eib", 1152921504606846976},   // 1024 * 1024 * 1024 * 1024 * 1024 * 1024
+
+        // SI units
+        {"kb", 1000},                    // 10e3
+        {"mb", 1000000},                 // 10e6
+        {"gb", 1000000000},              // 10e9
+        {"tb", 1000000000000},          // 10e12
+        {"pb", 1000000000000000},       // 10e15
+        {"eb", 1000000000000000000},    // 10e18
+    };
+    constexpr UInt64 MAX_UINT64 = std::numeric_limits<UInt64>::max();
+
+    class FunctionFromReadableSize : public IFunction
+    {
+    public:
+        static constexpr auto name = "fromReadableSize";
+        static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionFromReadableSize>(); }
+
+        String getName() const override { return name; }
+
+        bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+        size_t getNumberOfArguments() const override { return 1; }
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+        {
+            if (arguments.empty())
+                throw Exception(
+                    ErrorCodes::TOO_FEW_ARGUMENTS_FOR_FUNCTION,
+                    "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                    getName(),
+                    arguments.size());
+
+            if (arguments.size() > 1)
+                throw Exception(
+                    ErrorCodes::TOO_MANY_ARGUMENTS_FOR_FUNCTION,
+                    "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                    getName(),
+                    arguments.size());
+
+            const IDataType & type = *arguments[0];
+
+            if (!isString(type))
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Cannot format {} as time string.", type.getName());
+
+            return std::make_shared<DataTypeUInt64>();
+        }
+
+        bool useDefaultImplementationForConstants() const override { return true; }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+        {
+            auto col_to = ColumnUInt64::create();
+            auto & res_data = col_to->getData();
+
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                std::string_view str{arguments[0].column->getDataAt(i)};
+                Int64 token_tail = 0;
+                Int64 token_front = 0;
+                Int64 last_pos = str.length() - 1;
+                UInt64 result = 0;
+
+                /// ignore '.' and ' ' at the end of string
```

Comment:
Ignoring spaces and dots at the end of the string seems arbitrary. 1. One might wonder, why other trailing characters like `-` or enclosing single/double quotes or backticks are not ignored. 2. We allow patterns like `20 MiB . . . . . .`, that seems plain silly.

Honestly, I suggest to remove l. 98/99.

(an exception is acceptable if there is precedence for this kind of fault-tolerant parsing in other databases, please let me know. MySQL with its very creative function implementation comes to mind...)

Reply:
Sure, removed it, don't have a strong opinion here. I've now enforced the no-existence of leading/trailing whitespace, an error will be raised if this happens.

Would this be too strict?

Reply:
The way it is implemented after your latest push sounds right to me, thanks.